enhancements/cert-manager/istio-csr-controller.md

```diff
@@ -0,0 +1,929 @@
+	// +optional
+	TrustDomain string `json:"trustDomain,omitempty"`
+
+	// RootCAFile is for setting the file location containing the root CA which is
+	// present in the configured IssuerRef. File should be made available using the
+	// Volume and VolumeMount options.
+	// +kubebuilder:default:="/var/run/secrets/istio-csr/ca.crt"
+	// +optional
+	RootCAFile string `json:"rootCAFile,omitempty"`
+
+	// CertificateDNSNames contains DNS names to be added to the certificate SAN.
+	// +kubebuilder:example:=["cert-manager-istio-csr.istio-system.svc"]
+	// +optional
+	CertificateDNSNames []string `json:"certificateDNSNames,omitempty"`
+
+	// IstiodCertificateDuration is the istio's certificate validity period.
+	// +kubebuilder:default:="1h"
+	// +optional
+	IstiodCertificateDuration time.Duration `json:"istiodCertificateDuration,omitempty"`
+
+	// IstiodCertificateRenewBefore is the ahead time to renew the istio's certificate before
+	// expiry.
+	// +kubebuilder:default:="30m"
+	// +optional
+	IstiodCertificateRenewBefore time.Duration `json:"istiodCertificateRenewBefore,omitempty"`
+
+	// IstiodPrivateKeySize is the key size to be for RSAKey.
+	// +kubebuilder:default:=2048
+	// +optional
+	IstiodPrivateKeySize int32 `json:"istiodPrivateKeySize,omitempty"`
+}
+
+// ServerConfig is for configuring the server endpoint used by istio
+// for ontaining the certificates.
+type ServerConfig struct {
+	// Address to serve istio-csr gRPC service.
+	// +kubebuilder:default:="0.0.0.0"
+	// +optional
+	Address string `json:"address,omitempty"`
+
+	// Port to serve istio-csr gRPC service.
+	// +kubebuilder:default:="6443"
+	// +optional
+	Port string `json:"port,omitempty"`
+
+	// CertificateKeySize is the server's serving certificate's key size.
+	// +kubebuilder:default:=2048
+	// +optional
+	CertificateKeySize int32 `json:"certificateKeySize,omitempty"`
+
+	// SignatureAlgorithm is the signature algorithm to use when generating
+	// private keys. At present only RSA and ECDSA are supported.
+	// +kubebuilder:default:="RSA"
+	// +kubebuilder:validation:Enum:="RSA";"ECDSA"
+	// +optional
+	SignatureAlgorithm string `json:"signatureAlgorithm,omitempty"`
+
+	// MaxCertificateDuration is the maximum validity duration that can be
+	// requested for a certificate.
+	// +kubebuilder:default:="1h"
+	// +optional
+	MaxCertificateDuration time.Duration `json:"certificateDuration,omitempty"`
+}
+
+// IstioConfig is for configuring the istio specifics.
+type IstioConfig struct {
+	// Revisions are the istio revisions that are currently installed in the cluster.
+	// Changing this field will modify the DNS names that will be requested for
+	// the istiod certificate.
+	// +kubebuilder:default:=["basic"]
+	// +optional
+	Revisions []string `json:"revisions,omitempty"`
+
+	// namespace of the istio control-plane. In the same namespace issuer will be created
+	// used for obtaining the serving certificates.
+	Namespace string `json:"namespace,omitempty"`
```

Comment:
istio control-plane namespace should be mandatory.

Reply:
Will make it mandatory.

Reply:
What is the rbac configuration needed to install in the istio control plane namespace?

Reply:
Just for the reference, it was initially planned as markers, this will be added as manifest templates to create on demand.
```yaml
//+kubebuilder:rbac:groups="cert-manager.io",resources=clusterissuer,verbs=get;list;watch;create;delete
//+kubebuilder:rbac:groups="cert-manager.io",namespace=istio-system,resources=certificate;issuer,verbs=get;list;watch;create;delete
//+kubebuilder:rbac:groups="",resources=events;services;namespaces;serviceaccounts;configmaps,verbs=get;list;watch;create;update;patch;delete
//+kubebuilder:rbac:groups="rbac.authorization.k8s.io",resources=clusterroles;clusterrolebindings,verbs=get;list;watch;create;update;patch;delete
//+kubebuilder:rbac:groups="rbac.authorization.k8s.io",namespace=istio-system,resources=roles;rolebindings,verbs=get;list;watch;create;update;patch;delete
```